File: app/src/test/java/com/egecius/rxjava2_demo_2/rx/subjects/MaybeSubjectExamplesTest.java
package com.egecius.rxjava2_demo_2.rx.subjects;

import static com.egecius.rxjava2_demo_2.rx.utils.TestingUtils.setUncaughtExceptionsHandlerAsStrict;

import com.egecius.rxjava2_demo_2.rx.EgisException;
import com.egecius.rxjava2_demo_2.rx.utils.UncaughtException;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.observers.TestObserver;

@RunWith(MockitoJUnitRunner.class)
public class MaybeSubjectExamplesTest {

    private MaybeSubjectExamples mSut;

    @Before
    public void setUp() {
        setUncaughtExceptionsHandlerAsStrict();
        mSut = new MaybeSubjectExamples();
    }

    @Test
    public void succeedsWhenOnlyOneErrorEmitted() {
        TestObserver<String> testObserver = mSut.getMaybeSubject().test();

        mSut.emitSingleError();

        testObserver.assertError(EgisException.class);
    }


    @Test (expected = UncaughtException.class)
    public void succeedsWhen2ErrorEmitted() {
        TestObserver<String> testObserver = mSut.getMaybeSubject().test();

        mSut.emit2Errors();

        testObserver.assertError(EgisException.class);
    }

}

File: app/src/test/java/com/egecius/rxjava2_demo_2/rx/subjects/MaybeSubjectExamplesTest.kt
package com.egecius.rxjava2_demo_2.rx.subjects

import com.egecius.rxjava2_demo_2.rx.EgisException
import com.egecius.rxjava2_demo_2.rx.utils.TestingUtils.setUncaughtExceptionsHandlerAsStrict
import com.egecius.rxjava2_demo_2.rx.utils.UncaughtException
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class MaybeSubjectExamplesTest {

    private var mSut: MaybeSubjectExamples? = null

    @Before
    fun setUp() {
        setUncaughtExceptionsHandlerAsStrict()
        mSut = MaybeSubjectExamples()
    }

    @Test
    fun succeedsWhenOnlyOneErrorEmitted() {
        val testObserver = mSut!!.maybeSubject.test()

        mSut!!.emitSingleError()

        testObserver.assertError(EgisException::class.java)
    }


    @Test(expected = UncaughtException::class)
    fun succeedsWhen2ErrorEmitted() {
        val testObserver = mSut!!.maybeSubject.test()

        mSut!!.emit2Errors()

        testObserver.assertError(EgisException::class.java)
    }

}